src/main/jni/src/genomicsdb_GenomicsDBQuery.cc

```diff
@@ -0,0 +1,319 @@
+/**
+ * The MIT License (MIT)
+ * Copyright (c) 2019 Omics Data Automation, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of 
+ * this software and associated documentation files (the "Software"), to deal in 
+ * the Software without restriction, including without limitation the rights to 
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of 
+ * the Software, and to permit persons to whom the Software is furnished to do so, 
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all 
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS 
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER 
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN 
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "genomicsdb_GenomicsDBQuery.h"
+#include "genomicsdb.h"
+
+#include <mutex>
+
+bool is_initialized_ = false;
+std::mutex initializing_mutex_;
+
+//java.util.ArrayList
+jclass java_ArrayList_ ;
+jmethodID java_ArrayList_init_;
+jmethodID java_ArrayList_size_;
+jmethodID java_ArrayList_get_;
+jmethodID java_ArrayList_add_;
+
+//java.util.Hashmap
+jclass java_HashMap_;
+jmethodID java_HashMap_init_;
+jmethodID java_HashMap_put_;
+
+//org.genomicsdb.reader.GenomicsDBQuery$VariantCalls
+jclass java_VariantCalls_;
+jmethodID java_VariantCalls_init_default_;
+jmethodID java_VariantCalls_init_;
+
+//org.genomicsdb.reader.GenomicsDBQuery$VariantCall
+jclass java_VariantCall_;
+jmethodID java_VariantCall_init_;
+
+//org.genomicsdb.reader.GenomicsDBQuery$Pair
+jclass java_Pair_;
+jmethodID java_Pair_init_;
+jmethodID java_Pair_getStart_;
+jmethodID java_Pair_getEnd_;
+
+#define INIT(VAR,CODE)                                                                                 \
+  do {                                                                                                 \
+    VAR = CODE;                                                                                        \
+    if (VAR == NULL) {                                                                                 \
+      throw GenomicsDBException("genomicsdb_GenomicsDBQuery.cc#initialize:"+std::to_string(__LINE__)); \
+    }                                                                                                  \
+  } while (false)
+
+void initialize(JNIEnv *env) {
+  std::lock_guard<std::mutex> guard(initializing_mutex_);
```

Comment:
An easier way to  do this is to create a class called JNIInitializer (for example). Constructor does the initialization. Instantiate a single global object of this class. This way when the library is loaded, the initialization is done (no need to worry about race conditions etc).

Reply:
Yep, I was thinking of doing it in JNI_OnLoad, but did not want  to initialize when the api was not being invoked. On the other hand, maybe not having a mutex may be beneficial.

Reply:
@kgururaj , used Java_org_genomicsdb_reader_GenomicsDBQuery_jniInitialize() to initialize the JNI api interface in the static constructor as you suggested. Thanks.